src/object.hpp

```diff
@@ -178,20 +191,323 @@ template <typename Derived> class readable_object {
+            assert(obj_ != nullptr && index_ < size_);
+            return obj_[index_];
+        }
+
+        std::pair<object_key, object_view> operator*() const
+        {
+            assert(obj_ != nullptr && index_ < size_);
+            const auto &slot = obj_[index_];
+            if (type_ == object_type::map) {
+                return {&slot, slot};
+            }
+            return {{}, slot};
+        }
+
+        [[nodiscard]] std::size_t index() const { return static_cast<std::size_t>(index_); }
+
+        iterator &operator++() noexcept
+        {
+            // Saturated increment (to size)
+            index_ += static_cast<uint32_t>(index_ < size_);
+            return *this;
+        }
+
+        [[nodiscard]] iterator prev() const noexcept
+        {
+            // Saturated decrement (to 0)
+            auto new_index = static_cast<uint16_t>(index_ - static_cast<uint16_t>(index_ > 0));
+            return {obj_, size_, new_index, type_};
+        }
+
+    protected:
+        iterator() = default;
+
+        explicit iterator(const detail::object *obj, uint16_t idx = 0)
+            : obj_(obj->array), size_(static_cast<uint16_t>(obj->nbEntries)), index_(idx),
+              type_(static_cast<object_type>(obj->type))
+        {}
+
+        iterator(
+            // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+            const detail::object *obj, std::uint16_t size, std::uint16_t index, object_type type)
+            : obj_(obj), size_(size), index_(index), type_(type)
+        {}
+
+        const detail::object *obj_{nullptr};
+        std::uint16_t size_{0};
+        std::uint16_t index_{0};
+        object_type type_{object_type::invalid};
+
+        friend class object_view;
+    };
+
+    [[nodiscard]] iterator begin() const
+    {
+        assert(obj_ != nullptr);
+        // This check guarantees that the object is a container and not null
+        if (!is_container()) {
+            [[unlikely]] return {};
+        }
+        return iterator{obj_};
+    }
+
+    [[nodiscard]] iterator end() const
+    {
+        assert(obj_ != nullptr);
+        // This check guarantees that the object is a container and not null
+        if (!is_container()) {
+            [[unlikely]] return {};
+        }
+        return iterator{obj_, static_cast<uint16_t>(obj_->nbEntries)};
+    }
+
+protected:
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-const-or-ref-data-members)
+    const detail::object *obj_{nullptr};
+};
+
+static_assert(sizeof(object_view) == sizeof(void *));
 
 template <typename Derived> class writable_object {
 public:
     [[nodiscard]] borrowed_object at(std::size_t idx);
 
     borrowed_object emplace_back(owned_object &&value);
     borrowed_object emplace(std::string_view key, owned_object &&value);
+    borrowed_object emplace(owned_object &&key, owned_object &&value);
+
+    template <typename T>
+    borrowed_object emplace_back(const T &value)
+        requires(!std::is_same_v<T, owned_object>);
```

Comment:
why this requirement? I would expect that you take here a forwarding reference and require that that works for building an object inside the array. Anyway, that's how "emplace" generally works, you create the object in place there with the given constructor arguments